### ISSUE TITLE
"IDE0036: Modifiers are not ordered" - inconsistent ordering

### ISSUE THREAD

**Issue by SetTrend** [dotnet/roslyn#50738]
The "IDE0036: Modifiers are not ordered" rule evaluates modifier sequence inconsistently.

## Brief description:

It looks like this rule prefers `abstract` to _prefix_ the `internal` access modifier while it's preferring the same to _postfix_ the `protected` access modifier.

## Languages applicable:

I checked with C# analyzer only.

## Code example that the analyzer should report:

![Modifierts are not ordered](https://user-images.githubusercontent.com/9283914/105638606-c5efa900-5e73-11eb-9831-194ebe889409.png)

### This piece of code was entered:

```c#
protected abstract void BeforePropertyChange();
protected abstract void Undo();
internal abstract void Save();
```
The final line was flagged by Code Analyzer. It suggested the following modifier sequence:

```c#
protected abstract void BeforePropertyChange();
protected abstract void Undo();
abstract internal void Save();
```

**Comment by Youssef1313** [comment]
Based on the following code:

https://github.com/dotnet/roslyn/blob/bd32ad5c5f8202a4193b4f29ed472aec70017ab0/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/CodeStyle/CSharpCodeStyleOptions.cs#L194-L205

Both `protected` and `internal` have the same level, and they both come before `abstract`.

I also couldn't reproduce using Visual Studio 16.8.3:

![image](https://user-images.githubusercontent.com/31348972/105640137-fbe95900-5e84-11eb-8e11-0856a8b8af7d.png)

Note that this rule is configurable, can you make sure that your `.editorconfig` doesn't set `csharp_preferred_modifier_order` option?

**Comment by SetTrend** [comment]
Thanks, @Youssef1313, for replying.

Yes, you are right. I had Visual Studio 2019 create an `.editorconfig` file to add a custom rule.

The `.editorconfig` file created by Visual Studio 2019 contains the rule you mentioned:

```
#when this rule is set to a list of modifiers, prefer the specified ordering.
csharp_preferred_modifier_order = public,private,protected,readonly,abstract:suggestion
```

How come that the generated `.editorconfig` file differs from the originally configured Visual Studio 2019 behaviour? Shouldn't it fully reflect the original Visual Studio behaviour when being generated?

**Comment by Youssef1313** [comment]
@SetTrend When I tried to let VS generate the `.editorconfig` on its own, I found:

```
# Modifier preferences
csharp_prefer_static_local_function = true:suggestion
csharp_preferred_modifier_order = public,private,protected,internal,static,extern,new,virtual,abstract,sealed,override,readonly,unsafe,volatile,async:silent
```

Could you provide the steps that results in the missing `internal` modifier?

**Comment by SetTrend** [comment]
I finished my sample project now and uploaded it to Github. Please [find the corresponding repository here](/SetTrend/MVVM-Test).

Later on, I will provide further steps to reproduce ...

**Comment by SetTrend** [comment]
This is so strange:

Today I deleted the existing `.editorconfig` file from my solution and had it re-created by Visual Studio 2019 ... and the new `.editorconfig` file now looks completely different from the `.editorconfig` file created 4 days ago by the same Visual Studio.

I just checked in the latest file.

Does anybody have an explanation for this disturbing behaviour?

**Comment by Youssef1313** [comment]
@sharwell Is there any kind of Machine Learning that generates `.editorconfig` based on the codebase?

**Comment by sharwell** [comment]
There are several different commands that can be used to create a **.editorconfig** file, and they all do very different things.

### Create .editorconfig to match current Tools &rarr; Options...

![image](https://user-images.githubusercontent.com/1408396/105861422-8cb65680-5fa3-11eb-8f91-54657ba69980.png)

### Create (mostly) empty .editorconfig

#### With `root = true`

![image](https://user-images.githubusercontent.com/1408396/105862237-6b099f00-5fa4-11eb-9bc0-573d564e81fa.png)

#### Without `root = true`

![image](https://user-images.githubusercontent.com/1408396/105861818-f9315580-5fa3-11eb-98ae-e1a820241bb3.png)

### Create .editorconfig with partial content that doesn't seem to match anything specific

![image](https://user-images.githubusercontent.com/1408396/105862755-013dc500-5fa5-11eb-92cd-0fd6fcaf7571.png)

### Create .editorconfig using IntelliCode (machine learning)

![image](https://user-images.githubusercontent.com/1408396/105863412-ba040400-5fa5-11eb-93f0-32813621e726.png)

**Comment by SetTrend** [comment]
I've been using the latter approach.

If the outcome of the "New EditorConfig" file isn't deterministic, that's not a good idea.

It feels highly uncomfortable to report failure to the team with no-one being able to reproduce, not even I myself.

**Comment by Youssef1313** [comment]
It may be good to rename the menu item to "New EditoConfig (using Maching Learning)" or something that makes it clear. @sharwell What do you think?

**Comment by SetTrend** [comment]
Obviously, the ML generated `.editorConfig` file is only a snapshot, valid at a certain point in time; probably not even valid at all.

If the ML generated `.editorConfig` varies over time, it would be required to have Visual Studio automatically regenerate the file to match the forth growing VS project situation.

**Comment by sharwell** [comment]
> It may be good to rename the menu item ...

I thought the name used to be "New EditorConfig (IntelliCode)" but perhaps I'm mistaken. It's provided by the IntelliCode extension (different team), and I normally have that extension disabled so I'm not sure when it changed.

I would suggest filing feedback through [Developer Community](https://developercommunity.visualstudio.com/) if you have suggestions regarding that command, since those developers wouldn't be watching this repository.

**Comment by SetTrend** [comment]
Thanks to all of you for your valuable contribution.

I'm going to switch over then to Developer Community ...

Actually, there is already a thread existing at Developer Community, dealing with a very similar situation:

https://developercommunity.visualstudio.com/content/problem/689621/unable-to-disable-ai-assisted-intellisense.html

**Comment by SetTrend** [comment]
I now created the appropriate issue at Developer Community:

https://developercommunity2.visualstudio.com/t/Add-editorConfig-created-by-Machine-Le/1322730
